src/structures/ApplicationCommand.js

```diff
@@ -156,6 +156,90 @@ class ApplicationCommand extends Base {
     return this.manager.delete(this, this.guildId);
   }
 
+  /**
+   * Whether this command equals another command. It compares all properties, so for most operations
+   * it is advisable to just compare `command.id === command2.id` as it is much faster and is often
+   * what most users need.
+   * @param {ApplicationCommand|ApplicationCommandData|APIApplicationCommand} command The command to compare with
+   * @returns {boolean}
+   */
+  equals(command) {
+    // If given an id, check if the id matches
+    if (command.id && this.id !== command.id) return false;
+
+    // Check top level parameters
+    const commandType = typeof command.type === 'string' ? command.type : ApplicationCommandTypes[command.type];
+    if (
+      command.name !== this.name ||
+      ('description' in command && command.description !== this.description) ||
+      (commandType && commandType !== this.type) ||
+      // Future proof for options being nullable
+      // TODO: remove ?? 0 on each when nullable
+      (command.options?.length ?? 0) !== (this.options?.length ?? 0) ||
+      (command.defaultPermission ?? command.default_permission ?? true) !== this.defaultPermission
+    ) {
+      return false;
+    }
+
+    if (command.options) {
+      return this.constructor.optionsEqual(this.options, command.options);
+    }
+    return true;
+  }
+
+  /**
+   * Recursively checks that all options for an {@link ApplicationCommand} are equal to the provided options.
+   * In most cases it is better to compare using {@link ApplicationCommand#equals}
+   * @param {ApplicationCommandOptionData[]} existing The options on the existing command,
+   * should be {@link ApplicationCommand#options}
+   * @param {ApplicationCommandOptionData[]|APIApplicationCommandOption[]} options The options to compare against
+   * @returns {boolean}
+   */
+  static optionsEqual(existing, options) {
+    if (existing.length !== options.length) return false;
+    for (const option of existing) {
+      const foundOption = options.find(o => o.name === option.name);
```

Comment:
would the indexes be the same for both arrays?

could do
```js
return existing.every((option, idx) => this._optionEquals(option, options[idx]));
```
in any case an `every` call would probably be what i'd use here

Reply:
The order in which you send options does actually matter so two commands should not be considered equal if they have the same options in a different order

Reply:
IIRC, discord sometimes reorders things, and reordered options / choices do not actually get taken into account.
I used for...of for two reasons
1. return directly from the loop (its one less op too!)
2. make it clear that the loop stops as soon as it hits a non-equal option
Bonus 3. If for some reason any of the functions need to be async in the future, we don't need to rewrite the loop

Reply:
Discord does take order into account and deploys options in the exact order you provide them, so this change is important

Reply:
Hmm, it does take it into account, but not everything is deployed in the order provided (client side that is, API still has the correct order) E.g. subcommand groups are **always** displayed before subcommands on the same level. 

Its hard to say which behavior would be more confusing given that. I imagine most people store the command definition somewhere in their code, but if you were to reconstruct a command from the way the client displayed it (again, you could just fetch it to be sure its identical), it could be different but actually appear identical.

Not to mention that the way subcommands / options are displayed looks to be being changed soon.

Reply:
If the API still has the same order then you can probably compare the indexes just fine, since it would actually change something

Reply:
Why restrict this to relying on the index order being the same between both API and user definitions if we have a method that is more robust and reliable?

Reply:
The idea is to check index order and then use the method tho?

Reply:
> If for some reason any of the functions need to be async in the future, we don't need to rewrite the loop

i don't see a function like this becoming async any time soon 😄 

on another note, the inital idea behind suggesting using the index was simply just not to iterate through the second options for every single option

Reply:
But you can’t just check for the index which I think is what you’re suggesting right?

Reply:
> i don't see a function like this becoming async any time soon 😄

Absolutely not heh

Okay, so here's my thoughts
1. Provide an optional param, enforceOrdering that will enforce strict equality in array orders
2. To alleviate concern with iterations, iterate through the second set of options once, storing them in a map by name to be able to perform the very quick get based on the name.

Reply:
Why would you need a parameter for that? I don’t see a reason to leave that disabled tbf
Point 2 is fine by me tho